### PR TITLE
feat: add `DataFrame` and `LazyFrame` `explode` method

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added
- [x] Documented the changes

### CODE REVIEW

File: narwhals/_arrow/dataframe.py [hunk 1]
```diff
@@ -743,3 +744,78 @@ def unpivot(
         )
         # TODO(Unassigned): Even with promote_options="permissive", pyarrow does not
         # upcast numeric to non-numeric (e.g. string) datatypes
+
+    def explode(self: Self, columns: str | Sequence[str], *more_columns: str) -> Self:
```
Comment (FBruzzesi) [inline narwhals/_arrow/dataframe.py:748]: pyarrow has two paths:

- if nulls or empty lists are not present, then it is enough to:
  1. make sure element counts are same
  2. explode each array individually
- if nulls or empty lists are present, then these are ignore by `pc.list_parent_indices` and `pc.list_flatten`, which _is_ a problem. This implementation falls back to a python list both to flatten the array(s) and to create the corresponding indices .

After flattening, a new table is created by `take`-ing the indices of the non-flattened arrays and the flattened arrays.

File: narwhals/_pandas_like/dataframe.py [hunk 2]
```diff
@@ -937,3 +937,52 @@ def unpivot(
                 value_name=value_name if value_name is not None else "value",
             )
         )
+
+    def explode(self: Self, columns: str | Sequence[str], *more_columns: str) -> Self:
```
Comment (FBruzzesi) [inline narwhals/_pandas_like/dataframe.py:941]: If a single column is to be exploded, then we use the pandas native method. If multiple columns, the strategy is to explode the one column with the rest of the dataframe, and the other series individually and finally concatenating them back, plus sorting by original column names order

File: narwhals/dataframe.py [hunk 3]
```diff
@@ -4643,3 +4720,56 @@ def unpivot(
...
+        Notes:
+            It is possible to explode multiple columns only if these columns must have
+            matching element counts.
+
+        Arguments:
+            columns: Column names. The underlying columns being exploded must be of the `List` data type.
+            *more_columns: Additional names of columns to explode, specified as positional arguments.
+
+        Returns:
+            New DataFrame
```
Comment (FBruzzesi) [inline narwhals/dataframe.py:4736]:
```suggestion
            New LazyFrame
```

File: narwhals/_arrow/dataframe.py [hunk 4]
```diff
@@ -749,3 +750,78 @@ def unpivot(
...
+                [
+                    i
+                    for i, count in enumerate(counts.to_pylist())
+                    for _ in range(max(count or 1, 1))
+                ]
+            )
+
+            def explode_null_array(array: pa.ChunkedArray) -> pa.ChunkedArray:
+                exploded_values = []  # type: ignore[var-annotated]
+                for lst_element in array.to_pylist():
```
Comment (MarcoGorelli) [inline narwhals/_arrow/dataframe.py:806]: i might be missing something but this looks potentially very expensive

Comment (FBruzzesi) [reply narwhals/_arrow/dataframe.py:806]: It definitely is 😐 happy to raise for nullable cases for now and try to figure out a native alternative

Comment (FBruzzesi) [reply narwhals/_arrow/dataframe.py]: I did some improvements 🙌🏼 now there is only one `.to_pylist` and no loop even in the case of nulls or empty lists

**Comment by MarcoGorelli** [comment]
thanks - tbh i'm still not too keen on having to_pylist, should be just not support pyarrow and raise this as a feature request with them?

**Comment by FBruzzesi** [comment]
> thanks - tbh i'm still not too keen on having to_pylist, should be just not support pyarrow and raise this as a feature request with them?

Let me take a detour in the rabbit hole before a final decision 😇

**Comment by FBruzzesi** [comment]
@MarcoGorelli thoughts on the following?

The idea being, if consecutive parent indices are the same, their diff will be zero, therefore the counter of such index should increase, otherwise it should not. By how much you ask? the cumulative count starting from 0 (that's why of the -1 after the cumulative sum) - I will definitly add a comment if we decide to move forward with this 

```py
parent_indices = pc.list_parent_indices(native_frame[to_explode[0]])
diff = pc.pairwise_diff(parent_indices.combine_chunks())
is_change_point = pc.or_kleene(pc.equal(diff, 0), pc.is_null(diff))
indices = pc.subtract(
    pc.add(
        parent_indices,
        pc.cumulative_sum(is_change_point.cast(pa.int32()))
    ), 1
)

is_valid_index = pc.is_in(indices, value_set=parent_indices)
exploded_size = len(is_valid_index)

def flatten_func(array: pa.ChunkedArray) -> pa.ChunkedArray:
    dtype = array.type.value_type
    return pc.replace_with_mask(
        pa.array([None] * exploded_size, type=dtype),
        is_valid_index,
        pc.list_flatten(array).combine_chunks(),
    )
```

**Comment by MarcoGorelli** [comment]
nice - imma have to spend some time understanding this 😄 😳

**Comment by FBruzzesi** [comment]
> nice - imma have to spend some time understanding this 😄 😳

I deduce that my explanation was garbage. Let me _explode_ the previous comment

**Comment by MarcoGorelli** [comment]
Not at all, I just hadn't spent enough time on it , will take a look and make sense of it!

File: narwhals/_pandas_like/dataframe.py [hunk 5]
```diff
@@ -949,3 +949,52 @@ def unpivot(
...
+            if isinstance(columns, str)
+            else [*columns, *more_columns]
+        )
+        schema = self.collect_schema()
+        for col_to_explode in to_explode:
+            dtype = schema[col_to_explode]
+
+            if dtype != dtypes.List:
+                msg = f"`explode` operation not supported for dtype `{dtype}`"
+                raise InvalidOperationError(msg)
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:969]: can we also write which type we expect (List) in the error message?

**Review by MarcoGorelli — COMMENTED** [review]
OK thanks for explaining - sure, happy to try this out, maybe I could write a hypothesis test to stress test it

would also be open to merging for now without pyarrow (as the rest is quite uncontroversial) and then adding pyarrow later?

File: narwhals/_arrow/dataframe.py [hunk 6]
```diff
@@ -752,3 +753,70 @@ def unpivot(
...
+        if fast_path:
+            indices = pc.list_parent_indices(native_frame[to_explode[0]])
+            flatten_func = pc.list_flatten
+
+        else:
+            msg = (
+                "`DataFrame.explode` is not supported for pyarrow backend and column"
+                "containing null's or empty list elements"
+            )
+            raise NotImplementedError(msg)
```
Comment (MarcoGorelli) [inline narwhals/_arrow/dataframe.py:808]: do we want a value-dependent error to be raised? would you be opposed to raising completely for pyarrow and raising a feature request on their part?

Comment (FBruzzesi) [reply narwhals/_arrow/dataframe.py:808]: Oh, I see! I am fairly happy with both cases implementations as they are native.
But it's also ok to have a dedicated PR for it. I will adjust

Comment (FBruzzesi) [reply narwhals/_arrow/dataframe.py:808]: Maybe I should also check how pandas does it with pyarrow backend?

**Review by MarcoGorelli — APPROVED** [review]
thanks @FBruzzesi ! feel free to merge when ready

**Pull request merged**